src/test/rbd_mirror/test_ImageDeleter.cc

```diff
@@ -0,0 +1,210 @@
+#include "tools/rbd_mirror/ImageDeleter.h"
+#include "tools/rbd_mirror/ImageReplayer.h"
+#include "tools/rbd_mirror/Threads.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/internal.h"
+#include "librbd/Utils.h"
+
+#include "test/librados/test.h"
+#include "gtest/gtest.h"
+
+#include <thread>
+#include <chrono>
+
+#define GLOBAL_IMAGE_ID "global_image_id"
+
+#define dout_subsys ceph_subsys_rbd_mirror
+
+using rbd::mirror::RadosRef;
+
+void register_test_rbd_mirror_image_deleter() {
+}
+
+class TestImageDeleter : public ::testing::Test {
+public:
+  TestImageDeleter()
+  {
+    m_local_cluster = RadosRef(new librados::Rados());
+    EXPECT_EQ("", connect_cluster_pp(*(m_local_cluster.get())));
+
+    m_local_pool_name = get_temp_pool_name();
+    EXPECT_EQ(0, m_local_cluster->pool_create(m_local_pool_name.c_str()));
+    m_local_pool_id = m_local_cluster->pool_lookup(m_local_pool_name.c_str());
+    EXPECT_EQ(0, m_local_cluster->ioctx_create(m_local_pool_name.c_str(),
+					      m_local_ioctx));
+
+    EXPECT_EQ("", connect_cluster_pp(m_remote_cluster));
+
+    m_remote_pool_name = get_temp_pool_name();
+    EXPECT_EQ(0, m_remote_cluster.pool_create(m_remote_pool_name.c_str()));
+    m_remote_pool_id = m_remote_cluster.pool_lookup(m_remote_pool_name.c_str());
+    EXPECT_GE(m_remote_pool_id, 0);
+
+    EXPECT_EQ(0, m_remote_cluster.ioctx_create(m_remote_pool_name.c_str(),
+					       m_remote_ioctx));
+
+    m_image_name = get_temp_image_name();
+    uint64_t features = g_ceph_context->_conf->rbd_default_features;
+    features |= RBD_FEATURE_EXCLUSIVE_LOCK | RBD_FEATURE_JOURNALING;
+    int order = 0;
+    EXPECT_EQ(0, librbd::create(m_remote_ioctx, m_image_name.c_str(), 1 << 22,
+				false, features, &order, 0, 0));
+    m_remote_image_id = get_image_id(m_remote_ioctx, m_image_name);
+
+    m_threads = new rbd::mirror::Threads(reinterpret_cast<CephContext*>(
+      m_local_cluster->cct()));
+
+    m_deleter = new rbd::mirror::ImageDeleter(m_local_cluster,
+        m_threads->timer, &m_threads->timer_lock);
+
+  }
+
+  ~TestImageDeleter()
+  {
+    delete m_threads;
+    delete m_deleter;
+
+    EXPECT_EQ(0, m_remote_cluster.pool_delete(m_remote_pool_name.c_str()));
+    m_local_cluster->pool_delete(m_local_pool_name.c_str());
+  }
+
+  void start_stop_replayer(int seconds) {
+    rbd::mirror::ImageReplayer<> *replayer = new rbd::mirror::ImageReplayer<>(
+        m_threads,
+        m_local_cluster,
+        rbd::mirror::RadosRef(new librados::Rados(m_remote_ioctx)),
+        m_local_mirror_uuid, m_remote_mirror_uuid, m_local_ioctx.get_id(),
+        m_remote_pool_id, m_remote_image_id, GLOBAL_IMAGE_ID
+    );
+
+    rbd::mirror::ImageReplayer<>::BootstrapParams bootstap_params(m_image_name);
+    C_SaferCond cond;
+    replayer->start(&cond, &bootstap_params);
+    ASSERT_EQ(0, cond.wait());
+
+    m_local_image_id = replayer->get_local_image_id();
+    m_local_image_name = replayer->get_local_image_name();
+
+    if (seconds > 0) {
```

Comment:
yeah, this was a nasty hack that I used to ensure (or not) that I wouldn't stop image replayer while it was still holding the local temporary snapshot. As a side effect, I found that if I stop during image replayer sync, the interruption handler does not remove the temporary snapshot, and the ImageDeleter will not be able to delete the local image because it fails while trying to delete the temporary snapshot with an -EROFS error. 

What do you suggest to ensure that the the sync is finished?